resources/assets/lib/virtual-list/virtual-list.tsx

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+// This is a port of https://github.com/developerdizzle/react-virtual-list updated for React 18
+// using typescript and mobx with other unnecessary parts removed.
+
+import { throttle } from 'lodash';
+import { action, computed, makeObservable, observable } from 'mobx';
+import { observer } from 'mobx-react';
+import * as React from 'react';
+
+export interface Props<T> {
+  children: (props: RenderProps<T>) => React.ReactNode;
+  itemBuffer: number;
+  itemHeight: number;
+  items: T[];
+}
+
+export interface RenderProps<T> {
+  items: T[];
+}
+
+const emptyItemBounds = { firstItemIndex: 0, lastItemIndex: 0 };
+
+function topFromWindow(element: Window | HTMLElement | Element | null): number {
+  if (element == null) return 0;
+
+  const offsetTop = 'offsetTop' in element ? element.offsetTop : 0;
+  const offsetParent = 'offsetParent' in element ? element.offsetParent : null;
+
+  return offsetTop + topFromWindow(offsetParent);
+}
+
+
+@observer
+export default class VirtualList<T> extends React.Component<Props<T>> {
+  static readonly defaultProps = {
+    itemBuffer: 0,
+  };
+
+  private readonly ref = React.createRef<HTMLDivElement>();
+  private readonly scrollContainer = window; // we only care about window for now.
+  @observable private scrollTop = 0;
+  private readonly throttledSetScroll = throttle(() => this.setScroll(), 10);
```

Comment:
the 10ms throttle seems to be only useful on systems with high refresh rate 🤔 With enough item buffer, even 33ms is probably fine...?

Reply:
60fps is 16.6ms; it's also more noticeable when scrolling fast with the scroll wheel.
Larger buffer means more elements for React to render and if you scroll fast enough, everything gets repainted, anyway, so the buffer has minimal effect beyond scrolling a certain speed.
`scroll` events in current browsers are already supposed to be throttled at the same rate as requestAnimationFrame, anyway

Reply:
well, what I'm saying is the throttle isn't doing much (if at all) on 60hz displays. Even on 120hz it's still only throttling very little (10ms vs 8ms).

Reply:
The throttle is there to cap unlimited fps scenarios, otherwise it works fine without any throttle at all.